### PR TITLE
Stop copying LogicalPlan and Exprs in EliminateNestedUnion

### LINKED ISSUE #10296: Stop copying LogicalPlan and Exprs in `EliminateNestedUnion`
### Is your feature request related to a problem or challenge?
Part of https://github.com/apache/datafusion/issues/9637

As part of making the planner faster, we are updating the optimizer passes to avoid copying `LogicalPlan` and `Expr` (see https://github.com/apache/datafusion/issues/9637)

### Describe the solution you'd like

I would like to reduce the amount of copying in this pass (even though it doesn't appear in current profiling)

### Describe alternatives you've considered

Apply the model from @Lordworms  in https://github.com/apache/datafusion/pull/10166 to this pass
2. Update OptimizerRule::supports_rewrite` to return true
1. Update `OptimizerRule` to use `rewrite`
3. Update the pass itself to not copy the LogicalPlan (ideally using the [`TreeNode`](https://github.com/apache/datafusion/blob/b41ef20c5dad7bdd674e3cc5f35a9c99efae676c/datafusion/common/src/tree_node.rs#L33-L94) API) - it is implemented for `LogicalPlan` ([API](https://github.com/apache/datafusion/blob/b41ef20c5dad7bdd674e3cc5f35a9c99efae676c/datafusion/expr/src/logical_plan/tree_node.rs#L18-L38)) and `Expr` ([API](https://github.com/apache/datafusion/blob/b41ef20c5dad7bdd674e3cc5f35a9c99efae676c/datafusion/expr/src/tree_node.rs#L18)) 

Other examples: https://github.com/apache/datafusion/pull/10218

### Additional context

I think this will be a pretty simple refactor with a good existing pattern, so mark this as good first issue

### PR DESCRIPTION
**Pull request by emgeee**
## Which issue does this PR close?

Closes #10296.

## Are these changes tested?

Existing tests pass

### CODE REVIEW

**Comment by emgeee** [comment]
In addition to updating the interface for this optimizer, I tried to remove as many copies as possible but I'm still new to rust + datafusion so I'm very much open to suggestions and feedback!

File: datafusion/optimizer/src/eliminate_nested_union.rs [hunk 1]
```diff
@@ -37,49 +38,61 @@ impl EliminateNestedUnion {
 impl OptimizerRule for EliminateNestedUnion {
     fn try_optimize(
         &self,
-        plan: &LogicalPlan,
+        _plan: &LogicalPlan,
         _config: &dyn OptimizerConfig,
     ) -> Result<Option<LogicalPlan>> {
+        internal_err!("Should have called EliminateFilter::rewrite")
```
Comment (alamb) [inline datafusion/optimizer/src/eliminate_nested_union.rs:44]:
```suggestion
        internal_err!("Should have called EliminateNestedUnion::rewrite")
```

**Review by alamb — COMMENTED** [review]
Nice! Thank you @emgeee


I think this is a step forward for sure 


One thing I noticed is that

```rust
fn extract_plans_from_union(plan: &Arc<LogicalPlan>) -> Vec<Arc<LogicalPlan>> {
    match plan.as_ref() {
        LogicalPlan::Union(Union { inputs, schema }) => inputs
            .iter()
            .map(|plan| Arc::new(coerce_plan_expr_for_schema(plan, schema).unwrap()))
            .collect::<Vec<_>>(),
        _ => vec![plan.clone()],
    }
}
```

Calls [`coerce_plan_expr_for_schema`](https://github.com/apache/datafusion/blob/af3950698879a1e159fbe579bd433a0d5ce1c7cb/datafusion/expr/src/expr_rewriter/mod.rs#L208-L233) which looks like it may still clone the input and exprs. 

Maybe we can try to get rid of those clones too (but as a follow on PR)

**Comment by alamb** [comment]
Marking as draft as I think this PR is no longer waiting on feedback. Please mark it as ready for review when it is ready for another look

**Comment by emgeee** [comment]
I went ahead and fixed the formatting so all tests pass and managed to remove 1 clone() call from with in the `coerce_plan_expr_for_schema()` call stack.

If we want to optimize `coerce_plan_expr_for_schema` further, I'd definitely agree a separate PR makes sense as it is called from a number of other locations as well

**Review by alamb — APPROVED** [review]
Thanks you for this contribution @emgeee  -- a very nice first contribution :bowtie: .

> If we want to optimize coerce_plan_expr_for_schema further, I'd definitely agree a separate PR makes sense as it is called from a number of other locations as well

I agree -- this would be great to remove additional copies. If you had the time PRs would be most appreciated 🙏

**Comment by alamb** [comment]
Thanks again @emgeee

**Pull request merged**